src/ORM/DataList.php

```diff
@@ -308,72 +313,106 @@ public function distinct($value)
+            // If $sort is string then convert string to array to allow for validation
+            if (is_string($sort)) {
+                $newSort = [];
+                // Making the assumption here there are no commas in column names
+                // Other parts of silverstripe will break if there are commas in column names
+                foreach (explode(',', $sort) as $colDir) {
+                    if (empty($colDir)) {
+                        continue;
+                    }
+                    // Using regex instead of explode(' ') in case column name includes spaces
+                    if (preg_match('/^(.+) (asc|desc)$/i', trim($colDir), $matches)) {
+                        list($column, $direction) = [$matches[1], $matches[2]];
+                    } else {
+                        list($column, $direction) = [$colDir, 'ASC'];
+                    }
+                    $newSort[$column] = $direction;
+                }
+                $sort = $newSort;
+            }
+        }
+        // Validate that column names are not raw sql
+        if (is_array($sort)) {
+            foreach (array_keys($sort) as $column) {
+                $this->validateColumnHasNoRawSql($column);
+            }
         }
-
         return $this->alterDataQuery(function (DataQuery $query, DataList $list) use ($sort) {
-
             if (is_string($sort) && $sort) {
-                if (false !== stripos($sort ?? '', ' asc') || false !== stripos($sort ?? '', ' desc')) {
+                if (preg_match('/^(.+) (asc|desc)$/i', $sort)) {
                     $query->sort($sort);
                 } else {
                     $list->applyRelation($sort, $column, true);
                     $query->sort($column, 'ASC');
                 }
             } elseif (is_array($sort)) {
-                // sort(array('Name'=>'desc'));
-                $query->sort(null, null); // wipe the sort
-
+                // Wipe the sort
+                $query->sort(null, null);
                 foreach ($sort as $column => $direction) {
-                    // Convert column expressions to SQL fragment, while still allowing the passing of raw SQL
-                    // fragments.
                     $list->applyRelation($column, $relationColumn, true);
                     $query->sort($relationColumn, $direction, false);
                 }
             }
         });
     }
 
+    private function validateColumnHasNoRawSql(string $column): void
+    {
+        // Remove quotes around the column name
+        $col = trim(str_replace('"', '', $column));
+        // Useing regex because $this->canSortBy() failed on complex relations e.g. Team.Founder.Surname and
+        // also increased the number of db queries
+        // Search for non-alphanumeric characters in the column name rather than ^[^a-zA-Z0-9_]$ to account
+        // for column names that include macrons etc
+        if (preg_match('#[\(\)\[\]{}\+\-\*/=\:\?\\;|&\'"`<>^%, ]#', $col)) {
```

Comment:
```suggestion
        if (preg_match('#[()[\]{}+\-*/=:?\\;|&'"`<>^%,\s]#', $col)) {
```
Only `]`, `\`, and `-` (and special character groupings like `\s`) need to be escaped from `[]` blocks in regex.
Use `\s` instead of ` ` to avoid unexpected problems with other whitespace characters.

Reply:
Updated